gson/src/main/resources/META-INF/proguard/gson.pro

```diff
@@ -54,16 +55,19 @@
   <init>();
 }
 
-# Keep fields annotated with @SerializedName for classes which are present.
+# Keep fields annotated with @SerializedName for classes which are referenced.
 # If classes with fields annotated with @SerializedName have a no-args
-# constructor keep that as well.
+# constructor keep that as well. Based on
+# https://issuetracker.google.com/issues/150189783#comment11.
+# See also https://github.com/google/gson/pull/2420#discussion_r1241813541
+# for a more detailed explanation.
 -if class *
 -keepclasseswithmembers,allowobfuscation class <1> {
   @com.google.gson.annotations.SerializedName <fields>;
 }
 -if class * {
   @com.google.gson.annotations.SerializedName <fields>;
 }
--keepclassmembers,allowobfuscation class <1> {
+-keepclassmembers,allowobfuscation,allowoptimization class <1> {
```

Comment:
This rule is only covering the no-args constructor (`<init>()`), so here there is no issue with `allowoptimization`. For the rule above for the annotated fields it should not be there as otherwise R8 can do value propagation.